lib/src/main/scala/spinal/lib/tools/HDElkDiagramGen.scala

```diff
@@ -0,0 +1,599 @@
+    writer.close()
+  }
+}
+
+object HDElkDiagramGen {
+
+  /** Generating all diagrams */
+  def apply[T <: Component](rtl: SpinalReport[T]): Unit = {
+    val fileName = rtl.toplevelName + ".html"
+    val file = new File(fileName)
+    var writer = new FileWriter(file)
+    writer.write(
+      s"""<!DOCTYPE html>
+         |<html>
+         |<head>
+         |    <meta charset="UTF-8">
+         |    <title> RTL diagrams of ${rtl.toplevelName}</title>
+         |    <style>
+         |.buttons-container {
+         | display: flex;
+         | justify-content: center;
+         | margin-top:30px;
+         |}
+         |
+         |.buttons-container button {
+         | background-color: #4CAF50;
+         | color: white;
+         | border: none;
+         | padding:10px20px;
+         | margin:010px;
+         | cursor: pointer;
+         | border-radius:5px;
+         |}
+         |
+         |.center-title {
+         |  text-align: center;
+         |  font-size:24px;
+         |  font-weight: bold;
+         |  margin-top:20px;
+         | margin-bottom: 20px;
+         |}
+         |.goTop {
+         |	width: 50px;
+         |	height: 50px;
+         |	background-color: aquamarine;
+         |	font-size: 20px;
+         |	text-align: center;
+         |	line-height: 25px;
+         |	color: azure;
+         |	position: fixed;
+         |	bottom: 50px;
+         |	right: 50px;
+         |	display: none;
+         |}
+         |    </style>
+         |</head>
+         |<body>
+         |<button class="goTop" id="gotop">Go Top</button>
+         |<script>
+         |	var goTop=document.getElementById("gotop")
+         |	    window.onscroll=function(){
+         |		var jhlheight=document.documentElement.scrollTop||document.body.scrollTop
+         |		if(jhlheight>=300){
+         |		goTop.style.display="block"
+         |		}else{
+         |		goTop.style.display="none"
+         |   }
+         |	}
+         |	  goTop.onclick=function(){
+         |		window.scrollTo({
+         |			top:0,
+         |			behavior:"smooth"
+         |		})
+         |	}
+         |</script>
+         |<script src="https://cdn.jsdelivr.net/gh/Readon/hdelk@master/js/elk.bundled.js"></script>
+         |<script src="https://cdn.jsdelivr.net/gh/Readon/hdelk@master/js/svg.min.js"></script>
+         |<script src="https://cdn.jsdelivr.net/gh/Readon/hdelk@master/js/hdelk.js"></script>
+         |
+         |<h1 class="center-title">choose diagrams</h1>
+         |<div class="buttons-container">
+         |<a href="#${rtl.toplevelName}"><button>${rtl.toplevelName}</button></a>&nbsp;
+         |""".stripMargin
+    )
+    val module = rtl.toplevel
+    val allInnerCells = module.children
+    for (cell <- allInnerCells) {
+      writer.write(
+        s"""<a href="#${cell.getName()}"><button>${
+          cell
+            .getName()
+        }</button></a>&nbsp;\n""")
+    }
+    writer.write(s"""</div><br><br><br><br>\n""")
+    writer.close()
```

Comment:
See above note about not opening/closing the writer multiple times. It should really only be opened once and then passed as a parameter.